metaflow/plugins/unbounded_foreach_decorator.py

```diff
@@ -0,0 +1,135 @@
+
+        executable = self.environment.executable(step_name)
+        script = sys.argv[0]
+
+        # Access the `unbounded_foreach` param using `flow` (as datastore).
+        assert(flow._unbounded_foreach)
+        foreach_iter = flow.input
+        if not isinstance(foreach_iter, InternalUnboundedForeachInput):
+            raise MetaflowException('Expected type to be '\
+                                    'InternalUnboundedForeachInput. Found %s'\
+                                    % (type(foreach_iter)))
+        foreach_num_splits = sum(1 for _ in foreach_iter)
+
+        print('Simulating UnboundedForeach over value:',
+              foreach_iter, 'num_splits:', foreach_num_splits)
+        mapper_tasks = []
+
+        for i in range(foreach_num_splits):
+            task_id = \
+                '%s-%d' % (control_task_id.replace('control-', 'test-ubf-'), i)
+            pathspec = '%s/%s/%s' % (run_id, step_name, task_id)
+            mapper_tasks.append(to_unicode(pathspec))
+            input_paths = '%s/%s/%s' % (run_id, split_step_name, split_task_id)
+
+            # Override specific `step` kwargs.
+            kwargs = cli_args.step_kwargs
+            kwargs['split_index'] = str(i)
+            kwargs['run_id'] = run_id
+            kwargs['task_id'] = task_id
+            kwargs['input_paths'] = input_paths
+            kwargs['ubf_context'] = UBF_TASK
+            kwargs['retry_count'] = 0
+
+            cmd = cli_args.step_command(executable, script, step_name,
+                                        step_kwargs=kwargs)
+            step_cli = u' '.join(cmd)
+            # Print cmdline for execution. Doesn't work without the temporary
+            # unicode object while using `print`.
+            print(u'[${cwd}] Starting split#{split} with cmd:{cmd}'\
+                  .format(cwd=os.getcwd(),
+                          split=i,
+                          cmd=step_cli))
+            output_bytes = subprocess.check_output(cmd)
+            output = to_unicode(output_bytes)
+            for line in output.splitlines():
+                print('[Split#%d] %s' % (i, line))
+        # Save the list of (child) mapper task pathspec(s) into a designated
+        # artifact `_control_mapper_tasks`.
+        flow._control_mapper_tasks = mapper_tasks
+
+
+    def task_decorate(self,
+                      step_func,
+                      flow,
+                      graph,
+                      retry_count,
+                      max_user_code_retries,
+                      ubf_context):
+        if ubf_context == UBF_CONTROL:
+            from functools import partial
+            return partial(self.control_task_step_func, flow, graph, retry_count)
+        else:
+            return step_func
+
+    def step_task_retry_count(self):
+        # UBF plugins don't want retry for the control task. We signal this
+        # intent to the runtime by returning (None, None).
```

Comment:
It would require a bit more change and changing the existing decorators. I think this may be a good idea to revisit when we redo the whole API for decorators (which is hopefully going to happen at some point).